### PR TITLE
adding auto extraction validation checks for incident types

### PR DESCRIPTION
**Pull request by roysagi**
## Status
- [x] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
related: https://github.com/demisto/etc/issues/28221

## Description
adding auto extraction validation checks for incident types

### CODE REVIEW

File: demisto_sdk/commands/common/errors.py [hunk 1]
```diff
@@ -874,6 +875,22 @@ def incident_type_integer_field(field):
     def incident_type_invalid_playbook_id_field():
         return 'The "playbookId" field is not valid - please enter a non-UUID playbook ID.'
 
+    @staticmethod
+    @error_code_decorator
+    def incident_type_auto_extract_fields_invalid(incident_fields):
+        return f"The following incident fields are not formatted correctly under " \
+               f"`fieldCliNameToExtractSettings`: {incident_fields}\n" \
+               f"Please format them in open of the following ways:\n" \
```
Comment (Itay4) [inline demisto_sdk/commands/common/errors.py:883]: something in the english here doesnt work for me..

File: demisto_sdk/commands/common/errors.py [hunk 2]
```diff
@@ -874,6 +875,22 @@ def incident_type_integer_field(field):
...
+               f"1. To extract all indicators from the field: \n" \
+               f"isExtractingAllIndicatorTypes: true, extractAsIsIndicatorTypeId: \"\", " \
+               f"extractIndicatorTypesIDs: []\n" \
+               f"2. To extract the incident field to a specific indicator without using regex: \n" \
+               f"isExtractingAllIndicatorTypes: false, extractAsIsIndicatorTypeId: \"INDICATOR_TYPE\", " \
```
Comment (Itay4) [inline demisto_sdk/commands/common/errors.py:888]:
```suggestion
               f"isExtractingAllIndicatorTypes: false, extractAsIsIndicatorTypeId: \"<INDICATOR_TYPE>\", " \
```

File: demisto_sdk/commands/common/errors.py [hunk 3]
```diff
@@ -874,6 +875,22 @@ def incident_type_integer_field(field):
...
+               f"extractIndicatorTypesIDs: []\n" \
+               f"3. To extract indicators from the field using regex: \n" \
+               f"isExtractingAllIndicatorTypes: false, extractAsIsIndicatorTypeId: \"\", " \
+               f"extractIndicatorTypesIDs: [\"INDICATOR_TYPE1\", \"INDICATOR_TYPE2\"]"
```
Comment (Itay4) [inline demisto_sdk/commands/common/errors.py:892]:
```suggestion
               f"extractIndicatorTypesIDs: [\"<INDICATOR_TYPE1>\", \"<INDICATOR_TYPE2>\"]"
```

**Comment by Itay4** [comment]
@roysagi can you please open a draft pr in the content repo with incident type modified with the relevant changes to run a build with the changes in this pr?

File: demisto_sdk/commands/common/hook_validations/incident_type.py [hunk 4]
```diff
@@ -119,3 +120,39 @@ def is_valid_playbook_id(self):
             if self.handle_error(error_message, error_code, file_path=self.file_path):
                 return False
         return True
+
+    def is_valid_autoextract(self):
```
Comment (Itay4) [inline demisto_sdk/commands/common/hook_validations/incident_type.py:124]: please add docstring

File: demisto_sdk/commands/common/hook_validations/incident_type.py [hunk 5]
```diff
@@ -119,3 +120,39 @@ def is_valid_playbook_id(self):
...
+        auto_extract_data = self.current_file.get('extractSettings', {}).get('fieldCliNameToExtractSettings')
+
+        # no auto extraction set in incident type.
+        if not auto_extract_data:
+            return True
+
+        invalid_incident_fields = []
+        for incident_field in auto_extract_data:
```
Comment (Itay4) [inline demisto_sdk/commands/common/hook_validations/incident_type.py:132]: i think its better to do something like:
```python
for incident_field, extract_setting in auto_extract_data.items():
    extracting_all = extract_setting.get('isExtractingAllIndicatorTypes')
```

File: demisto_sdk/commands/common/hook_validations/incident_type.py [hunk 6]
```diff
@@ -119,3 +120,39 @@ def is_valid_playbook_id(self):
...
+            extract_as_is = auto_extract_data[incident_field].get('extractAsIsIndicatorTypeId')
+            extracted_indicator_types = auto_extract_data[incident_field].get('extractIndicatorTypesIDs')
+
+            # General format check.
+            if type(extracting_all) != bool or type(extract_as_is) != str or type(extracted_indicator_types) != list:
+                invalid_incident_fields.append(incident_field)
+
+            # If trying to extract without regex make sure extract all is set to
+            # False and the extracted indicators list is empty
+            elif extract_as_is != "":
```
Comment (Itay4) [inline demisto_sdk/commands/common/hook_validations/incident_type.py:143]:
```suggestion
            elif extract_as_is:
```

File: demisto_sdk/commands/common/hook_validations/incident_type.py [hunk 7]
```diff
@@ -119,3 +120,39 @@ def is_valid_playbook_id(self):
...
+                if extracting_all is True or len(extracted_indicator_types) > 0:
```
Comment (Itay4) [inline demisto_sdk/commands/common/hook_validations/incident_type.py:144]:
```suggestion
                if extracting_all or extracted_indicator_type:
```

File: demisto_sdk/commands/common/hook_validations/incident_type.py [hunk 8]
```diff
@@ -119,3 +120,39 @@ def is_valid_playbook_id(self):
...
+                    invalid_incident_fields.append(incident_field)
+
+            # If trying to extract with regex make sure extract all is set to
+            # False and the extract_as_is should be set to an empty string
+            elif len(extracted_indicator_types) > 0:
```
Comment (Itay4) [inline demisto_sdk/commands/common/hook_validations/incident_type.py:149]:
```suggestion
            elif extracted_indicator_types:
```

File: demisto_sdk/commands/common/hook_validations/incident_type.py [hunk 9]
```diff
@@ -119,3 +120,39 @@ def is_valid_playbook_id(self):
...
+                if extracting_all is True or extract_as_is != "":
```
Comment (Itay4) [inline demisto_sdk/commands/common/hook_validations/incident_type.py:150]:
```suggestion
                if extracting_all or extract_as_is:
```

File: demisto_sdk/commands/common/errors.py (hunk 1 above)
Comment (roysagi) [reply demisto_sdk/commands/common/errors.py]: typo - fixing

File: demisto_sdk/commands/common/errors.py (hunk 2 above)
Comment (roysagi) [reply demisto_sdk/commands/common/errors.py]: done

File: demisto_sdk/commands/common/errors.py (hunk 3 above)
Comment (roysagi) [reply demisto_sdk/commands/common/errors.py]: done

File: demisto_sdk/commands/common/hook_validations/incident_type.py (hunk 4 above)
Comment (roysagi) [reply demisto_sdk/commands/common/hook_validations/incident_type.py:124]: done

File: demisto_sdk/commands/common/hook_validations/incident_type.py (hunk 5 above)
Comment (roysagi) [reply demisto_sdk/commands/common/hook_validations/incident_type.py]: done - good point

File: demisto_sdk/commands/common/hook_validations/incident_type.py (hunk 6 above)
Comment (roysagi) [reply demisto_sdk/commands/common/hook_validations/incident_type.py:148]: I kinda wanna be explicit here in order for this to be more understanding about what is being checked - even if it looks kinda redundant.

File: demisto_sdk/commands/common/hook_validations/incident_type.py (hunk 7 above)
Comment (roysagi) [reply demisto_sdk/commands/common/hook_validations/incident_type.py:149]: same

File: demisto_sdk/commands/common/hook_validations/incident_type.py (hunk 8 above)
Comment (roysagi) [reply demisto_sdk/commands/common/hook_validations/incident_type.py:154]: same

File: demisto_sdk/commands/common/hook_validations/incident_type.py (hunk 9 above)
Comment (roysagi) [reply demisto_sdk/commands/common/hook_validations/incident_type.py:155]: same

**Review by Itay4 — APPROVED** [review]

**Comment by roysagi** [comment]
Moving to on hold - waiting for the feature to be finalized.

**Comment by roysagi** [comment]
@Itay4 - as requested here is a PR with all possible forms of incident types: https://github.com/demisto/content/pull/10667
1) with extraction set to all and mode to all
2) extraction set to specific fields and mode set to all
3) extraction set to specific fields and mode set to specific
4) extraction set to all and mode set to specific
5) legacy - no extraction fields

**Review by Itay4 — APPROVED** [review]
excellent

**Pull request merged**